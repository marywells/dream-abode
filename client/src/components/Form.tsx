import { useState } from 'react';
import { IProperty } from '../types';

interface IProps {
  addNewProperty: (property: IProperty) => void;
}
export function Form({ addNewProperty }: IProps) {
  const [name, setName] = useState('');
  const [imageURL, setImageURL] = useState('');
  const [description, setDescription] = useState('');
  const [price, setPrice] = useState(0);
  const [beds, setBeds] = useState(0);
  const [baths, setBaths] = useState(0);
  const [type, setType] = useState('');

  function createProperty(newProperty: IProperty) {
    addNewProperty(newProperty);
  }

  function handleSubmit(event: React.FormEvent<HTMLFormElement>) {
    event.preventDefault();
    const property = {
      name,
      imageURL,
      description,
      price,
      beds,
      baths,
      type,
    };

    createProperty(property);

    setName('');
    setImageURL('');
    setDescription('');
    setPrice(0);
    setBeds(0);
    setBaths(0);
    setType('');
  }

  return (
    <div className={styles.form}>
      <h1 className={styles.formHeader}>Add a Property</h1>
      <form onSubmit={handleSubmit}>
        <div className={styles.inputLabel}>Name</div>
        <input
          className={styles.input}
          type='text'
          placeholder='First line of address'
          value={name}
          onChange={(e) => setName(e.target.value)}
        />
        <div className={styles.inputLabel}>Image</div>
        <input
          className={styles.input}
          type='text'
          placeholder='Image URL'
          value={imageURL}
          onChange={(e) => setImageURL(e.target.value)}
        />
        <div className={styles.inputLabel}>Description</div>
        <input
          className={styles.input}
          type='text'
          placeholder='Short description'
          value={description}
          onChange={(e) => setDescription(e.target.value)}
        />
        <div className={styles.inputLabel}>Price</div>
        <input
          className={styles.input}
          type='number'
          placeholder='Set price'
          value={price}
          onChange={(e) => setPrice(e.target.valueAsNumber)}
        />
        <div className={styles.inputLabel}>Bedrooms</div>
        <input
          className={styles.input}
          type='number'
          placeholder='Set bedrooms'
          value={beds}
          onChange={(e) => setBeds(e.target.valueAsNumber)}
        />
        <div className={styles.inputLabel}>Bathrooms</div>
        <input
          className={styles.input}
          type='number'
          placeholder='Set bathrooms'
          value={baths}
          onChange={(e) => setBaths(e.target.valueAsNumber)}
        />
        <div className={styles.inputLabel}>Property Type</div>
        <select
          className={styles.input}
          value={type}
          onChange={(e) => setType(e.target.value)}
        >
          <option value='Terraced'>Terraced</option>
          <option value='Flat'>Flat</option>
          <option value='Detached'>Detached</option>
          <option value='Detached'>Semi-Detached</option>
          <option value='Castle'>Castle</option>
        </select>
        <br />
        <input className={styles.button} type='submit' value='Create' />
      </form>
    </div>
  );
}

const styles = {
  form: 'w-72 md:fixed md:top-20 md:right-2 bg-gray-800 border p-4 m-2 rounded-lg',
  formHeader: 'text-white text-xl font-bold ',
  inputLabel: 'text-white text-xs pt-2',
  input: 'rounded text-sm w-full',
  button:
    'bg-transparent text-white font-semibold hover:text-white border mt-2 py-1 px-3 rounded-full',
};
