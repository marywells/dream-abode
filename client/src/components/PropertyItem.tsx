import { IProperty } from '../types';

export function PropertyItem(props: IProperty) {
  const { name, type, description, imageURL, price, beds, baths } = props;
  return (
    <div className={styles.propertyContainer}>
      <div data-testid='propertyItemName' className={styles.nameTile}>
        <p className={styles.name}>{name}</p>
      </div>

      <div className={styles.propertyDetails}>
        <div data-testid='propertyItemPrice' className={styles.price}>
          £{price?.toLocaleString('en-US')}
        </div>
        <div className={styles.type}>{type}</div>
        <div data-testid='propertyItemBeds'>{beds} bedrooms</div>
        <div>{baths} bathrooms</div>
        <br />
      </div>
      <div className={styles.imageContainer}>
        <img className={styles.image} src={imageURL} alt={name} />
      </div>
      <div className={styles.description}>{description}</div>
      <div className={styles.editDelete}>
        <button>✏️</button>
        <button>❌</button>
      </div>
    </div>
  );
}

const styles = {
  propertyContainer:
    'flex-wrap relative w-full border-2 rounded-lg shadow m-6 ml-2 text-xs flex',
  nameTile: 'flex bg-gray-900 rounded-lg items-center text-center',
  name: 'w-40 text-lg p-2 text-white ',
  imageContainer: 'w-40',
  image: 'h-28 rounded-lg',
  propertyDetails: 'w-24 pl-4 pt-2 mr-2',
  price: 'text-lg font-bold',
  type: 'font-semibold',
  description: 'hidden md:block w-72 p-2',
  editDelete: 'absolute bottom-2 right-2',
};
