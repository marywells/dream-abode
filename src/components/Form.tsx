export function Form() {
  return <div>Add a Property Form</div>;
}

const styles = {};
