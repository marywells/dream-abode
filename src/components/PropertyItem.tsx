import { IProperty } from '../types';

export function PropertyItem(props: IProperty) {
  return (
    <div className={styles.propertyContainer}>
      <p>{props.name}</p>
      <p>{props.type}</p>
      <p>{props.description}</p>
      <p>Â£{props.price}</p>
      <p>{props.beds} bedrooms</p>
      <p>{props.baths} bathrooms</p>
    </div>
  );
}

const styles = {
  propertyContainer: 'border-2 rounded-lg text-xs shadow m-2 p-1',
};
