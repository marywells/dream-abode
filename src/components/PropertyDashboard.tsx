import { useState, useEffect } from 'react';
import { PropertyItem } from './PropertyItem';
import { Form } from './Form';
import * as ApiService from '../api-service';
import { IProperty } from '../types';

export function PropertyDashboard() {
  const [properties, setProperties] = useState<IProperty[]>([]);

  useEffect(() => {
    loadProperties();
  }, []);

  async function loadProperties() {
    await ApiService.getAll().then((data) => {
      setProperties(data);
    });
  }

  function addNewProperty(property: IProperty) {
    ApiService.postOne(property).then(() => loadProperties());
  }

  return (
    <div className={styles.dashboardContainer}>
      <div className={styles.propertyList}>
        <h1>Property Dashboard</h1>
        {properties.map((property) => {
          return (
            <div key={property._id}>
              {
                <PropertyItem
                  _id={property._id}
                  name={property.name}
                  imageURL={property.imageURL}
                  description={property.description}
                  price={property.price}
                  beds={property.beds}
                  baths={property.baths}
                  type={property.type}
                />
              }
            </div>
          );
        })}
      </div>

      <Form addNewProperty={addNewProperty} />
    </div>
  );
}

const styles = {
  dashboardContainer: 'relative pt-16',
  propertyList: 'absolute flex-col',
};
