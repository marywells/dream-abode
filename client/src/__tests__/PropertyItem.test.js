import ReactDOM from 'react-dom';
import { PropertyItem } from '../components/PropertyItem';
import '@testing-library/jest-dom/extend-expect';
import { render, cleanup } from '@testing-library/react';

const mock = {
  name: '1 Test Street',
  description: 'A lovely street',
  imageURL: 'www.test.com',
  price: '200000',
  beds: 4,
  baths: 2,
  type: 'Flat',
};

afterEach(cleanup);
it('renders property item on the page', () => {
  const div = document.createElement('div');
  ReactDOM.render(<PropertyItem />, div);
});

it('renders property item data', () => {
  const { getByTestId } = render(
    <PropertyItem name={mock.name} price={mock.price} beds={mock.beds} />
  );
  expect(getByTestId('propertyItemName')).toHaveTextContent('1 Test Street');
  expect(getByTestId('propertyItemPrice')).toHaveTextContent('Â£200000');
  expect(getByTestId('propertyItemBeds')).toHaveTextContent('4 bedrooms');
});

it('matches property item snapshot', () => {
  const tree = renderer
    .create(
      <PropertyItem
        name={mock.name}
        description={mock.description}
        imageURL={mock.imageURL}
        price={mock.price}
        beds={mock.beds}
        baths={mock.baths}
        type={mock.type}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
