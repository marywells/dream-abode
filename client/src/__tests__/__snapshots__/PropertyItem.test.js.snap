// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches property item snapshot 1`] = `
<div
  className="flex-wrap relative w-full border-2 rounded-lg shadow m-6 ml-2 text-xs flex"
>
  <div
    className="flex bg-gray-900 rounded-lg items-center text-center"
    data-testid="propertyItemName"
  >
    <p
      className="w-40 text-lg p-2 text-white "
    >
      1 Test Street
    </p>
  </div>
  <div
    className="w-24 pl-4 pt-2 mr-2"
  >
    <div
      className="text-lg font-bold"
      data-testid="propertyItemPrice"
    >
      £
      200000
    </div>
    <div
      className="font-semibold"
    >
      Flat
    </div>
    <div
      data-testid="propertyItemBeds"
    >
      4
       bedrooms
    </div>
    <div>
      2
       bathrooms
    </div>
    <br />
  </div>
  <div
    className="w-40"
  >
    <img
      alt="1 Test Street"
      className="h-28 rounded-lg"
      src="www.test.com"
    />
  </div>
  <div
    className="hidden md:block w-72 p-2"
  >
    A lovely street
  </div>
  <div
    className="absolute bottom-2 right-2"
  >
    <button>
      ✏️
    </button>
    <button>
      ❌
    </button>
  </div>
</div>
`;
